### PR TITLE
Open Banking Sandbox demo integration into test env

### PR DESCRIPTION
**Pull request by rmarins**
### SUMMARY
Base implementation of the Open Banking Sandbox demo developed by the FSI BU.

##### ISSUE TYPE
- New role Pull Request

##### COMPONENT NAME
Role: ocp4-workload-openbanking-sandbox

##### ADDITIONAL INFORMATION
RITM0748851: "I resolved the lxml issue. The personal branch needs to be merged into the agnosticD DEV branch (default branch) and then we can test."

ansible 2.9.11
  config file = /home/rmarins/Workspaces/agnosticd/ansible.cfg
  configured module search path = ['/home/rmarins/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.8/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.8.5 (default, Aug 12 2020, 00:00:00) [GCC 10.2.1 20200723 (Red Hat 10.2.1-1)]

ansible==2.9.11
appdirs==1.4.3
argcomplete==1.10.0
attrs==19.3.0
awscli==1.17.17
Babel==2.8.0
bcrypt==3.1.7
Beaker==1.10.0
beautifulsoup4==4.9.1
blivet==3.2.2
blivet-gui==2.1.15
boto3==1.14.38
botocore==1.17.38
Brlapi==0.7.0
cachetools==3.1.1
certifi==2018.10.15
cffi==1.14.0
chardet==3.0.4
chrome-gnome-shell==0.0.0
cliff==2.16.0
cmd2==0.9.16
colorama==0.4.1
configobj==5.0.6
cryptography==2.8
cupshelpers==1.0
dasbus==0.2
dbus-python==1.2.16
debtcollector==1.22.0
decorator==4.4.0
defusedxml==0.6.0
distlib==0.3.1
distro==1.4.0
dnspython==2.0.0
docutils==0.15.2
dogpile.cache==0.9.0
filelock==3.0.12
fluidity-sm==0.2.0
fros==1.1
funcsigs==1.0.2
google-auth==1.20.0
gpg==1.14.0
humanize==0.5.1
idna==2.8
invoke==1.4.1
iso8601==0.1.12
jeepney==0.4.3
Jinja2==2.11.2
jmespath==0.9.4
jsonpatch==1.21
jsonpointer==1.10
jsonschema==3.2.0
k8s==0.16.0
keyring==21.3.0
keystoneauth1==3.17.1
kubernetes==11.0.0
langtable==0.0.52
lexicon==1.0.0
libcomps==0.1.15
lxml==4.4.1
Mako==1.1.1.dev0
MarkupSafe==1.1.1
msgpack==0.6.2
munch==2.3.2
netaddr==0.7.19
netifaces==0.10.6
nftables==0.1
ntlm-auth==1.1.0
ntplib==0.3.3
oauthlib==3.0.2
olefile==0.46
openshift==0.11.2
openstacksdk==0.36.0
ordered-set==3.1
os-client-config==1.33.0
os-service-types==1.7.0
osc-lib==1.14.1
oslo.config==6.11.1
oslo.context==2.23.0
oslo.i18n==3.24.0
oslo.log==3.44.1
oslo.serialization==2.29.2
oslo.utils==3.41.1
paramiko==2.7.1
Paste==3.2.4
pbr==5.4.3
pid==2.2.3
Pillow==7.0.0
ply==3.11
prettytable==0.7.2
productmd==1.27
psutil==5.6.7
pwquality==1.4.2
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycairo==1.18.2
pycparser==2.19
pycrypto==2.6.1
pycups==2.0.1
pycurl==7.43.0.5
pyenchant==2.0.0
PyGObject==3.36.1
pyinotify==0.9.6
PyJWT==1.7.1
pykickstart==3.24
PyNaCl==1.3.0
pyOpenSSL==19.0.0
pyparsing==2.4.7
pyparted==3.11.4
pyperclip==1.6.4
pyRFC3339==1.1
pyrsistent==0.16.0
PySocks==1.7.1
python-augeas==0.5.0
python-cinderclient==5.0.0
python-dateutil==2.8.0
python-glanceclient==2.17.0
python-keystoneclient==3.21.0
python-meh==0.48
python-neutronclient==6.14.0
python-novaclient==15.1.0
python-openstackclient==4.0.0
python-string-utils==1.0.0
python3-openid==3.1.0
pytz==2020.1
pyudev==0.22.0
pywinrm==0.3.0
pyxdg==0.26
PyYAML==5.3.1
requests==2.22.0
requests-file==1.4.3
requests-ftp==0.3.1
requests-ntlm==1.1.0
requests-oauthlib==1.2.0
requestsexceptions==1.4.0
rfc3986==1.2.0
rpm==4.15.1
rsa==3.4.2
ruamel.yaml==0.16.10
ruamel.yaml.clib==0.2.0
s3transfer==0.3.3
SecretStorage==3.1.1
selinux==3.0
sepolicy==3.0
setools==4.3.0
simplejson==3.17.0
simpleline==1.6
six==1.12.0
slip==0.6.4
slip.dbus==0.6.4
sos==3.9
soupsieve==1.9.2
stevedore==1.31.0
systemd-python==234
Tempita==0.5.1
terminator==1.92
udica==0.2.3
urllib3==1.25.7
virtualenv==20.0.30
warlock==1.3.3
wcwidth==0.2.4
websocket-client==0.56.0
wrapt==1.11.2
xmltodict==0.12.0

### CODE REVIEW

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-layout.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    _layout_content: "{{ lookup('url', _layout_location, split_lines=False) }}"
+
+- name: Update pre-exising Layout "{{ _layout_system_name }}"
+  when:
+    - _xpath_layouts is defined
+    - _xpath_layouts.matches is defined
+    - _xpath_layouts.matches | length | int > 0
+  block:
+
+  - set_fact: _layout_id="{{ _xpath_layouts.matches[0].id }}"
```
Comment (jkupferer) [inline ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-layout.yml:33]: Please use consistent YAML list indentation. The `when` clause above is indenting list items and the block here is not. Either way is fine so long as it is consistent.

Also please add a `.yamllint` file in your role directory. (ex: https://github.com/redhat-cop/agnosticd/blob/development/ansible/roles/bookbag/.yamllint)

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-layout.yml [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+  - uri:
```
Comment (jkupferer) [inline ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-layout.yml:35]: All tasks should include `name`.

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-layout.yml [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      body_format: form-urlencoded
+      body:
+      - [ draft, '{{ _layout_content }}' ]
+      status_code: 200
+
+- name: Upload a new Layout "{{ _layout_system_name }}"
+  when: (_xpath_layouts is undefined) or (_xpath_layouts.matches is undefined)
+  block:
+
+  - uri:
```
Comment (jkupferer) [inline ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-layout.yml:50]: All tasks should include `name`

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-page.yml [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      - [ path, '{{ _page_path }}' ]
+      - [ layout_name, '{{ _page_layout }}' ]
+      - [ type, 'page' ]
+      - [ liquid_enabled, '1' ]
+      - [ draft, '{{ _page_content }}' ]
+      return_content: yes
+      status_code: 201
+    register: _page_created
+
+  - xml:
```
Comment (jkupferer) [inline ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-page.yml:104]: All tasks should have `name`

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-page.yml [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+      xmlstring: '{{ _page_created.content }}'
+      xpath: "//page/id"
+      content: text
+    register: _created_page_id
+
+  - set_fact: _page_id="{{ _created_page_id.matches[0].id }}"
```
Comment (jkupferer) [inline ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-page.yml:110]: `name`

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/microcks-import-collection.yml [hunk 6]
```diff
@@ -0,0 +1,29 @@
+---
+# create temp file
+- tempfile:
```
Comment (jkupferer) [inline ansible/roles/ocp4-workload-openbanking-sandbox/tasks/microcks-import-collection.yml:3]: task `name`

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/post_workload.yml [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+  register: _microcks_keycloak_secret
+  retries: 3
+  delay: 5
+  until: _microcks_keycloak_secret.resources|length > 0
+
+- set_fact:
+    _keycloak_admin_usr: "{{ _microcks_keycloak_secret.resources[0].data.username | b64decode }}"
+    _keycloak_admin_pwd: "{{ _microcks_keycloak_secret.resources[0].data.password | b64decode }}"
+
+# - name: Authenticating Keycloak user "{{ _keycloak_admin_usr }}" via OAuth2
```
Comment (jkupferer) [inline ansible/roles/ocp4-workload-openbanking-sandbox/tasks/post_workload.yml:36]: Disabled code blocks should not be added to version control.

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/post_workload.yml [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+#     msg: "TOKEN ---> {{ _microcks_serviceaccount_token }}"
+
+
+
+
+
+
+
+
+
```
Comment (jkupferer) [inline ansible/roles/ocp4-workload-openbanking-sandbox/tasks/post_workload.yml:143]: Too many blank lines. Please use `yamllint` to catch these issues.

**Review by jkupferer — DISMISSED** [review]

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-deploy-full-api.yml [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+- name: Create the ActiveDocs for 3scale product "{{ _api_resource_id }} ({{ _product_id }})"
+  when: ( _api_doc_id | int ) == 0
+  shell: >
+    curl -k -f -s -S -X POST '{{ _3scale_admin_url }}/admin/api/active_docs.json' \
+    --header 'Authorization: {{ _3scale_basic_auth_token }}' \
+    --form 'service_id={{ _product_id }}' \
+    --form 'system_name={{ _api_base_system_name }}_docs' \
+    --form 'name={{ _product_api_name }}' \
+    --form 'published=true' \
+    --form 'body=@{{ _tmp_apispec.path }}'
```
[the remark this answers is not in the document]
Comment (rmarins) [reply ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-deploy-full-api.yml:251]: yep, it was not posible to use `uri` module with 3scale CMS backend -- the 
now commented out (see commits up to `e74e328fd2d313d21b1d10138b5dbee3d2d6c2ab`)

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-deploy-full-api.yml [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+  when: ( _api_doc_id | int ) > 0
+  shell: >
+    curl -k -f -s -S -X PUT '{{ _3scale_admin_url }}/admin/api/active_docs/{{ _api_doc_id }}.json' \
+    --header 'Authorization: {{ _3scale_basic_auth_token }}' \
+    --form 'id={{ _api_doc_id }}' \
+    --form 'service_id={{ _product_id }}' \
+    --form 'system_name={{ _api_base_system_name }}_docs' \
+    --form 'name={{ _product_api_name }}' \
+    --form 'published=true' \
+    --form 'body=@{{ _tmp_apispec.path }}'
```
[the remark this answers is not in the document]
Comment (rmarins) [reply ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-deploy-full-api.yml:272]: resolved as above

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-layout.yml (hunk 1 above)
Comment (rmarins) [reply ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-layout.yml]: yaml lint added and all checkstyle cleared (see commits up to `e74e328fd2d313d21b1d10138b5dbee3d2d6c2ab`)

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-layout.yml (hunk 2 above)
Comment (rmarins) [reply ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-layout.yml]: resolved (sett commits up to `e74e328fd2d313d21b1d10138b5dbee3d2d6c2ab`)

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-layout.yml (hunk 3 above)
Comment (rmarins) [reply ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-layout.yml]: resolved

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-page.yml (hunk 4 above)
Comment (rmarins) [reply ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-page.yml]: resolved

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-page.yml (hunk 5 above)
Comment (rmarins) [reply ansible/roles/ocp4-workload-openbanking-sandbox/tasks/3amp-portal-upload-page.yml]: resolved

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/microcks-import-collection.yml (hunk 6 above)
Comment (rmarins) [reply ansible/roles/ocp4-workload-openbanking-sandbox/tasks/microcks-import-collection.yml]: resolved

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/post_workload.yml (hunk 7 above)
Comment (rmarins) [reply ansible/roles/ocp4-workload-openbanking-sandbox/tasks/post_workload.yml]: code block with comments removed

File: ansible/roles/ocp4-workload-openbanking-sandbox/tasks/post_workload.yml (hunk 8 above)
Comment (rmarins) [reply ansible/roles/ocp4-workload-openbanking-sandbox/tasks/post_workload.yml]: resolved

**Comment by rmarins** [comment]
Please, merge this PR based on last commit `e74e328fd2d313d21b1d10138b5dbee3d2d6c2ab`

**Comment by amayagil** [comment]
@jkupferer could you please validate and merge if all ok? Thanks!

**Comment by amayagil** [comment]
please @wkulhanek or @stencell could you merge @rmarins' PR? we need that in order to continue onboarding

**Review by stencell — APPROVED** [review]
All issues from @jkupferer addressed. Merging this to move the specific workload on.

**Pull request merged**